File: Behavioral/Iterator/Cpp/SongIterator.cpp
// SongIterator.cpp

#include "SongIterator.h"

SongIterator::SongIterator(const std::vector<Song>& songs) : songs(songs) {}

bool SongIterator::hasNext() {
  return position < songs.size();
}

Song SongIterator::next() {
  return hasNext() ? songs[position++] : Song("", "");
}

File: Behavioral/Iterator/CSharp/SongIterator.cs
// SongIterator.cs

using System.Collections.Generic;

public class SongIterator : IIterator<Song> {
  private readonly List<Song> _songs;
  private int _position = 0;

  public SongIterator(List<Song> songs) {
    _songs = songs;
  }

  public bool HasNext() {
    return _position < _songs.Count;
  }

  public Song? Next() {
    return HasNext() ? _songs[_position++] : null;
  }
}
